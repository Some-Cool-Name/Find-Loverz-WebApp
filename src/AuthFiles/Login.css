div {
    display: block;
}
body {
    margin: 0;
    font-family: Arial, Helvetica, sans-serif;
}
.header{
    height: 45px;
    overflow: hidden;
    background:linear-gradient(to right,#c471ed,#b621fe , #12c2e9); 
}

h3{
  margin-top: auto;
  margin-bottom: opx;
    align-items: baseline;
    color: white;
    font-family:cursive;
    font-size: xx-large;
}
.formLogin{
    flex-direction: column;
    width: 300px;
    padding: 5;
    margin: auto;
    margin-top: 100px;
    height: 420px;
    background: #ffffff;
    box-shadow: 0px 14px 80px rgba(34, 35, 58, 0.2);
    padding: 40px 55px 45px 55px;
    border-radius: 25px;
    transition: all .3s;
}
hr{
    margin-top: 0px;
    margin-bottom: 40px;
    background: linear-gradient(to right,#c471ed,#b621fe , #12c2e9);
}
.input-boxes{
    padding: 0%;
}

label.fieldDescription{
    box-sizing: border-box;
    text-align: start;
    width: 100%;
    padding: 1px;
    padding-left: 15px;
    padding-bottom: 1px;
    margin-inline-start: auto;
    margin-inline-end: auto;
}
input{
    box-sizing: border-box;
    width: 100%;
    height: calc(3em + 2px);
    margin: 0 0 1em;
    padding: 1em;
    border: 1px solid #ccc;
    border-radius: 1.5em;
    background: transparent;
    resize: none;
    outline: none;
    top: 0;
    left: 0;
    
}

input[type="text"]:focus, input[type="password"]:focus {
    border-color: #00bafa;
}
.button{
    cursor: pointer;
    background: linear-gradient(to right,#c471ed,#b621fe , #12c2e9);
    box-shadow: 1px 15px 20px 15px #840fe440;
    padding: 10px;
    font-size: 1em;
    border-radius: 25px;
    border: none;
    align-items: center;
    text-align: center;
    width: 160px;
    margin: 20px 0px;
    height: fit-content;
    text-decoration: none;
}
a.button{
    height: fit-content;
    color: white;  
}




