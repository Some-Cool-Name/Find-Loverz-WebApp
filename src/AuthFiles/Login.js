import React, { useState } from 'react';
import 'regenerator-runtime/runtime';
//import './Login.css';
import { Link, useHistory } from 'react-router-dom';
import { Formik } from 'formik';
import { globalStyles } from '../globalStyles';
import { loginValidate } from './UserValidation';
import { loginRequest } from '../BackendRequests/Authentication';
import { saveToStorage } from '../HelperClasses/StorageHandler';
import isLoggedIn from '../HelperClasses/LoginChecker';

function Login({ setUser, user }) {
    let history = useHistory();
    const [incorrectDetails, setIncorrectDetails] = useState(false);
    const [error, setError] = useState(false);
    const incorrectDetailsString = "Incorrect username or password";
    const unknownError = "Oops, something went wrong. Please try again later";


    const loginUser = async (values) => {
        // validate user input
        try {
            const result = await loginRequest(values); 
            if(result.message === "success"){
                // extract user info from result
                const objArray = [];
                Object.keys(result).forEach(key => objArray.push({
                    name: key,
                    rating: result[key]
                }));
                const loggedInUser = [objArray[0].rating[0]];
                // login logistics
                saveToStorage('user', loggedInUser);
                setUser(loggedInUser);
                history.push("/feed"); 
            }
            else if(result.message === "Wrong Password" || result.message === "error"){
                // incorrect details
                setIncorrectDetails(true);
                setUser(false)
                 
            } 
        } catch (error) {
            // something went wrong
            setError(true);
        }   
    }


    // checks if user is logged and moves user to feed if logged in.
    if(isLoggedIn()) history.push('/feed');
    
    return (
        
        <div className='container'>
            <div className='header'>
                <h3 className='appName'>find loverz</h3>
             </div>
            <div  className='formLogin'>
                
                <Formik    
                    initialValues={{
                    email: '', 
                    password: '',
                    }}
                    validationSchema={loginValidate}
                    onSubmit={(values, actions) => {
                    actions.resetForm();
                    loginUser(values);
                    }}
                >
                    {
                        props => (
                            <div style={inputDivStyle}  >
                                <h1>Log in</h1>
                                <br />
                                <hr width="60px;"  size="8"></hr>
                                <br />
                                <label className='fieldDescription'><b>Email</b></label>
                                    
                                    <input
                                        className='inputBox'
                                        type="text" 
                                        placeholder="Email..." 
                                        style={globalStyles.inputStyle}
                                        onChange={props.handleChange('email')}
                                        onBlur={props.handleBlur('email')}
                                        value={props.values.email}                                                                  
                                    />
                                <p
                                    style={globalStyles.errorText}>
                                    {props.touched.email && props.errors.email}
                                </p>
                                 <br />
                                    <label className='fieldDescription'><b>Password</b></label>
                                    <input 
                                        type="password"
                                        required 
                                        placeholder="Password..." 
                                        style={globalStyles.inputStyle}
                                        onChange={props.handleChange('password')} 
                                        onBlur={props.handleBlur('password')}
                                        value={props.values.password}
                                        alt='Placeholder'
                                    />
                                <p
                                    style={globalStyles.errorText}>
                                    {props.touched.password && props.errors.password}
                                    {incorrectDetails && incorrectDetailsString}
                                    {error && unknownError}
                                </p>
                                 <br /> 
                                <Link className='button' to="/feed" style={globalStyles.button} onClick={props.handleSubmit} ><b>Login</b> </Link>
                                <Link to="/signup" >register</Link>
                            </div>
                        )
                    }
                </Formik>
            </div>
        </div> 
    )

}
const inputDivStyle = {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
}
export default Login
