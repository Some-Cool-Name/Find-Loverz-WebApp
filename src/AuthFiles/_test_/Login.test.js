import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import {rest} from 'msw';
import {setupServer} from 'msw/node';
import { BrowserRouter } from "react-router-dom"
import Login from '../Login';
import userEvent from '@testing-library/user-event';
import { act, isElementOfType } from "react-dom/test-utils";
import renderer from 'react-test-renderer';



const server = setupServer(
    rest.get('https://lamp.ms.wits.ac.za/home/s1851427/loverzlog.php', (req, res, ctx) => {
      // Respond with a mocked user token that gets persisted
      // in the `sessionStorage` by the `Login` component.
      return res( ctx.status(200), ctx.json(
          {"login":[
              {"username":"demo",
              "name":"demo",
              "Birthday":"10-05-1993",
              "gender":"Male",
              "Sexuality":"Female"
            }],
            "success":"1",
            "message":"success"}))
    }),
  )
  
  // Enable API mocking before tests.
  beforeAll(() => server.listen())
  
  // Reset any runtime request handlers we may add during the tests.
  afterEach(() => server.resetHandlers())
  
  // Disable API mocking after the tests are done.
  afterAll(() => server.close())

it('renders correctly', () => {
    const tree = renderer.create(<BrowserRouter>
        <Login/>
    </BrowserRouter>).toJSON();
    expect(tree).toMatchSnapshot();
});

// given, when , then 
it('Login Inputs( email, password ) : Empty Fields Testing , Should return error ', () => {

    // render or load login screen : now we have login screen displayed
    render( 
        <BrowserRouter>
            <Login/>
        </BrowserRouter>
    );
    
    const email = screen.getByPlaceholderText('Email...'); 
    userEvent.type(email, '');
    expect(email).toHaveValue('');


    const password = screen.getByPlaceholderText('Password...'); 
    userEvent.type(password, '');
    expect(password).toHaveValue('');

    const loginBtn = screen.getByText('Login');
    userEvent.click(loginBtn);

    expect(loginBtn).toBeTruthy();

});

it('Login Inputs( email, password ) : invalid email and pasword , Should return error ', async () => {

    server.use(
        rest.get('https://lamp.ms.wits.ac.za/home/s1851427/loverzlog.php', (req, res, ctx) => {
          // Respond with "400 Internal Server Error" status for this test.
          
          return res(
            ctx.status(400),
            ctx.json({ message: 'Invalid details' }),
          )
        }),
      );
    // render or load login screen : now we have login screen displayed
    render( 
        <BrowserRouter>
            <Login/>
        </BrowserRouter>
    );
    
    const email = screen.getByPlaceholderText('Email...'); 
    userEvent.type(email, 'Martha');
    expect(email).toHaveValue('Martha');

    const password = screen.getByPlaceholderText('Password...'); 
    userEvent.type(password, '12345678');
    expect(password).toHaveValue('12345678');

    const loginBtn = screen.getByText('Login');
    userEvent.click(loginBtn);

    expect(loginBtn).toBeTruthy();

});
  

it('Login Inputs( email, password ) : valid email and pasword , Should go to feed page ', () => {

    // render or load login screen : now we have login screen displayed
    render( 
        <BrowserRouter>
            <Login/>
        </BrowserRouter>
    );
    
    const email = screen.getByPlaceholderText('Email...'); 
    userEvent.type(email, 'demo');
    expect(email).toHaveValue('demo');


    const password = screen.getByPlaceholderText('Password...'); 
    userEvent.type(password, 'demo');
    expect(password).toHaveValue('demo');

    const loginBtn = screen.getByText('Login');
    userEvent.click(loginBtn);

    expect(loginBtn).toBeTruthy();

});