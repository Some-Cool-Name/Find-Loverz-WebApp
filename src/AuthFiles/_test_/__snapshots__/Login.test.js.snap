// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <div
    className="container"
  >
    <div
      className="header"
    >
      <h3
        className="appName"
      >
        find loverz
      </h3>
    </div>
    <div
      className="formLogin"
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "justifyContent": "center",
          }
        }
      >
        <h1>
          Log In
        </h1>
        <br />
        <hr
          size="8"
          width="60px;"
        />
        <br />
        <label
          className="fieldDescription"
        >
          <b>
            Email
          </b>
        </label>
        <input
          className="inputBox"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Email..."
          style={
            Object {
              "marginTop": "10px",
            }
          }
          type="text"
          value=""
        />
        <p
          style={
            Object {
              "color": "crimson",
              "fontWeight": "bold",
              "marginBottom": 10,
              "marginTop": 6,
              "textAlign": "center",
            }
          }
        />
        <br />
        <label
          className="fieldDescription"
        >
          <b>
            Password
          </b>
        </label>
        <input
          alt="Placeholder"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Password..."
          required={true}
          style={
            Object {
              "marginTop": "10px",
            }
          }
          type="password"
          value=""
        />
        <p
          style={
            Object {
              "color": "crimson",
              "fontWeight": "bold",
              "marginBottom": 10,
              "marginTop": 6,
              "textAlign": "center",
            }
          }
        />
        <br />
        <a
          className="button"
          href="/feed"
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "grey",
              "color": "black",
              "marginTop": 10,
              "padding": 5,
              "textDecoration": "none",
            }
          }
        >
          <b>
            Login
          </b>
           
        </a>
        <a
          href="/signup"
          onClick={[Function]}
        >
          register
        </a>
      </div>
    </div>
  </div>
</div>
`;
